### PR TITLE
(RUN): initial run configuration

### PR DESCRIPTION
**Pull request by matklad**
Simple run config. I want to do two more things here:

* add integration test against a real cargo
* add GUI designer UI

### CODE REVIEW

**Comment by matklad** [comment]
@atsky @alexeykudinkin this now includes a test, so I'd like to merge it, and deal with proper UI separately.

File: src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+abstract class CargoTestCaseBase : PlatformTestCase() {
+    abstract val testDataPath: String
+
+    private val sdkPath: String? get() = RustSdkType.INSTANCE.suggestHomePath()
+
+    override fun runTest() {
+        if (sdkPath == null) {
+            System.err?.println("SKIP $name: no Rust sdk found")
+            return
+        }
```
Comment (matklad) [inline src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt]: Is there a better way to conditionally skip a test with JUnit 3?

File: src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            null
+        else
+            SdkConfigurationUtil.createAndAddSDK(sdkPath, RustSdkType.INSTANCE)!!
+
+    override fun setUp() {
+        super.setUp()
+        val data = LocalFileSystem.getInstance().findFileByPath(testDataPath)
+            ?: throw RuntimeException("No such directory: $testDataPath")
+
+        data.refresh(/* asynchronous = */ false, /* recursive = */ true)
```
Comment (matklad) [inline src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt]: This is fishy, but without this `data` includes some strange files which should not be there.

File: src/main/kotlin/org/rust/cargo/runconfig/RustApplicationConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        configurationModule.checkForWarning()
+    }
+
+    override fun getConfigurationEditor(): SettingsEditor<out RunConfiguration> =
+        RustRunConfigurationEditorForm()
+
+    override fun getState(executor: Executor, environment: ExecutionEnvironment): RunProfileState  {
+        val module = configurationModule.module!!
+        val moduleManager = ModuleRootManager.getInstance(module)
+        val sdk = moduleManager.sdk!!
```
Comment (matklad) [inline src/main/kotlin/org/rust/cargo/runconfig/RustApplicationConfiguration.kt]: Am I correct that `getState` is called only if `checkConfiguration` finished successfully? If this is not the case, then these `!!` can fire NPE.

Comment (alexeykudinkin) [inline src/main/kotlin/org/rust/cargo/runconfig/RustApplicationConfiguration.kt]: Can the docs verify this?

File: src/main/kotlin/org/rust/cargo/runconfig/RustRunner.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package org.rust.cargo.runconfig
+
+import com.intellij.execution.configurations.RunProfile
+import com.intellij.execution.executors.DefaultRunExecutor
+import com.intellij.execution.runners.DefaultProgramRunner
+
+class RustRunner: DefaultProgramRunner() {
```
Comment (alexeykudinkin) [inline src/main/kotlin/org/rust/cargo/runconfig/RustRunner.kt]: Missing spaces.

File: src/main/kotlin/org/rust/cargo/runconfig/RustApplicationConfiguration.kt (hunk 3 above)
Comment (matklad) [inline src/main/kotlin/org/rust/cargo/runconfig/RustApplicationConfiguration.kt]: Ah, they say just the opposite. I should return `null` if it is impossible to create a `State`. Thanks!

**Comment by matklad** [comment]
rebased

File: src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    override fun runTest() {
+        if (sdkPath == null) {
+            System.err?.println("SKIP $name: no Rust sdk found")
+            return
+        }
+        super.runTest()
+    }
+
+    override fun getTestProjectJdk(): Sdk? =
+        if (sdkPath == null)
```
Comment (alexeykudinkin) [inline src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt]: Why not `let` it?

File: src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt (hunk 1 above)
Comment (alexeykudinkin) [inline src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt]: That's ok for the smoke test.

File: src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt (hunk 5 above)
Comment (matklad) [inline src/test/kotlin/org/rust/cargo/CargoTestCaseBase.kt]: Yeah, `let` is cleaner here, fixed. I wish Rust-style `try!` was possible in Kotlin, so as to write

```Kotlin
SdkConfigurationUtil.createAndAddSDK(try!(sdkPath), RustSdkType.INSTANCE)!!
```

The nearest Kotlin alternative looks ugly:

```Kotlin
SdkConfigurationUtil.createAndAddSDK(sdkPath ?: return null, RustSdkType.INSTANCE)!!
```

**Comment by matklad** [comment]
Added a GUI designer based form.

The configuration is changed from `cargo run` to a generic cargo command configuration. We would deferentially need all sorts of specific configurations created from context in the future, but for know I think that providing maximum flexibility is a good thing.

**Comment by fischmax** [comment]
I might just be making a stupid mistake, but I can't get this to actually run. I set up a new project with `cargo new myProject --bin` and imported it into IntelliJ (15.03 Community on Windows 10) by clicking on the Cargo.toml as par the guide. In the project settings window (Ctrl+Alt+Shift+S) I selected the path to my Rust installation `C:\Program Files\Rust nightly GNU 1.8\` and I created a run config which just says `cargo run`. Typing this thing into a Windows cmd.exe or the IntelliJ terminal (Alt+F12) works without a flaw. But trying to execute the run config gives me

```
Information:cargo: cargo build
Information:19.02.2016 19:25 - Compilation completed with 1 error and 0 warnings in 453ms
Error:Module 'myProject' production: java.lang.NoSuchMethodError: kotlin.sequences.SequencesKt.generateSequence(Lkotlin/jvm/functions/Function0;)Lkotlin/sequences/Sequence;
```

I have a hunch that I probably misconfigured something somewhere but I really don't know where to look …

**Comment by matklad** [comment]
@fischmax it's definitely not your fault. I suspect that if you use the [15.0.4 EAP](https://confluence.jetbrains.com/display/IDEADEV/IDEA+15+EAP) then it should work. 

However, this was supposed to work for 15.0.3 as well =/

**Comment by matklad** [comment]
@fischmax are you getting the same error if you just "make" your project (Ctrl + F9)?

**Comment by fischmax** [comment]
Thanks for the fast reply ;-)

Yes, I do get the same error when making the project, with the shortcut, from the context menu and from the top menu.

**Comment by fischmax** [comment]
@matklad Updating Kotlin proofed more challenging than you would think, because Windows wouldn't let IntelliJ delete a folder in „C:\Program Files (x86)\…“ but after I got that sorted out it now works like a charm!

Thanks again for your help, this rocks (-:

**Comment by matklad** [comment]
>Updating Kotlin proofed more challenging than you would think

Hah, I'm using [nixos](http://nixos.org/) and so my IntelliJ IDEA is placed on a read only file system, so I know the pain :)

**Pull request merged**